File: examples/python_binding/machine_split_field.py
"""AutoSplit + 3-machine fleet demo, with both 2D and map views.

Large irregular field auto-splits into sub-fields (longer AABB side
larger than max_side triggers a bisect). The same 3-machine
Block+ByLength plan runs on each sub-field; each physical machine's
tour concatenates its per-sub-field work.

One sub-field OWNS the midline headland; the other reaches the split
line directly and uses the neighbour's turn space.

Logs to `maptrax_split_field_py`.
"""

import rerun as rr

import maptrax

from _common import (
    BIG_IRREGULAR_FIELD,
    DATUM,
    arc_polyline,
    arc_polyline_geo,
    log_machine_views,
    machine_color,
    polygon_geo,
    segment_length,
    swath_line,
    swath_line_geo,
)


FLEET_SIZE = 3
MAX_SIDE = 520.0
SWATH_WIDTH = 8.0
HEADLAND_COUNT = 2


def ring_polyline(ring):
    points = [[float(x), float(y)] for x, y in ring["points"]]
    if points and points[0] != points[-1]:
        points.append(points[0])
    return points


def ring_polyline_geo(planner, ring):
    closed_xy = list(ring["points"])
    if closed_xy and closed_xy[0] != closed_xy[-1]:
        closed_xy.append(closed_xy[0])
    latlon = planner.enu_to_wgs_batch([(float(x), float(y)) for x, y in closed_xy])
    return [[lat, lon] for (lat, lon) in latlon]


def main():
    rr.init("maptrax_split_field_py", spawn=True)

    planner = maptrax.Maptrax()
    planner.set_field(BIG_IRREGULAR_FIELD, DATUM)

    rr.log(
        "enu/field/border",
        rr.LineStrips2D([BIG_IRREGULAR_FIELD + [BIG_IRREGULAR_FIELD[0]]]),
    )
    rr.log(
        "geo/field/border",
        rr.GeoLineStrings(lat_lon=[polygon_geo(planner, BIG_IRREGULAR_FIELD)]),
    )

    part_count = planner.plan_field(
        swath_width=SWATH_WIDTH,
        angle_degrees=20.0,
        headland_count=HEADLAND_COUNT,
        decomposition=f"auto_split:{MAX_SIDE}",
    )

    print(f"=== AutoSplit {{ max_side: {MAX_SIDE} }} + {FLEET_SIZE}-machine fleet ===")
    print("   Field extent: ~950m x ~495m (irregular 8-vertex polygon)")
    print(f"   Split into {part_count} sub-field(s).\n")

    ring_color = (120, 120, 160)
    for part_index in range(part_count):
        part = planner.get_part(part_index)
        for ring_index, ring in enumerate(part["headlands"]):
            rr.log(
                f"enu/parts/part_{part_index}/headland_{ring_index}",
                rr.LineStrips2D([ring_polyline(ring)], colors=[ring_color]),
            )
            rr.log(
                f"geo/parts/part_{part_index}/headland_{ring_index}",
                rr.GeoLineStrings(
                    lat_lon=[ring_polyline_geo(planner, ring)],
                    colors=[ring_color],
                ),
            )
        non_owned = part["non_owned_splits"]
        if non_owned:
            print(
                f"   part {part_index} is NON-OWNER of {len(non_owned)} split(s)"
            )
        else:
            print(f"   part {part_index} OWNS its splits (midline headland inset fully)")

    all_plans = planner.plan_machines_all_parts(
        machines=FLEET_SIZE,
        pattern="block",
        balance="length",
        routing_strategy="greedy_nearest",
        local_improvement_passes=1,
        turn_model="reeds_shepp",
        min_turning_radius=3.0,
        swath_width=SWATH_WIDTH,
        machine_length=6.0,
        machine_width=3.0,
    )

    per_machine_swaths = [[] for _ in range(FLEET_SIZE)]
    per_machine_tour = [[] for _ in range(FLEET_SIZE)]
    per_machine_headlands = [[] for _ in range(FLEET_SIZE)]

    for part_index, part_plan in enumerate(all_plans):
        for machine in part_plan["machines"]:
            idx = machine["machine_index"]
            per_machine_swaths[idx].extend(machine["assigned_swaths"])
            per_machine_tour[idx].extend(machine["tour"])
            per_machine_headlands[idx].extend(machine["assigned_headland_arcs"])

            color = machine_color(idx)
            if machine["assigned_swaths"]:
                rr.log(
                    f"enu/parts/part_{part_index}/m{idx}/swaths",
                    rr.LineStrips2D(
                        [swath_line(s) for s in machine["assigned_swaths"]],
                        colors=[color],
                    ),
                )
                rr.log(
                    f"geo/parts/part_{part_index}/m{idx}/swaths",
                    rr.GeoLineStrings(
                        lat_lon=[swath_line_geo(planner, s) for s in machine["assigned_swaths"]],
                        colors=[color] * len(machine["assigned_swaths"]),
                    ),
                )

    print()
    print(
        "   {:<8} {:>8}  {:>10}  {:>10}".format(
            "machine", "swaths", "work_len", "work_s"
        )
    )
    makespan = 0.0
    total_work = 0.0
    for idx in range(FLEET_SIZE):
        work_len = sum(
            segment_length(s["points"][0], s["points"][-1])
            for s in per_machine_swaths[idx]
        )
        work_s = work_len
        print(
            "   {:<8} {:>8}  {:>10.1f}  {:>10.1f}".format(
                f"m{idx}", len(per_machine_swaths[idx]), work_len, work_s
            )
        )
        makespan = max(makespan, work_s)
        total_work += work_len

        fake_machine = {
            "assigned_swaths": per_machine_swaths[idx],
            "tour": per_machine_tour[idx],
            "assigned_headland_arcs": per_machine_headlands[idx],
        }
        log_machine_views(rr, planner, f"machines/m{idx}", fake_machine, machine_color(idx))

    print(f"\n   Total work length: {total_work:.1f} m")
    print(
        f"   Makespan: {makespan:.1f}s   ({FLEET_SIZE} machines across {part_count} sub-field(s))"
    )


if __name__ == "__main__":
    main()

File: examples/machine_split_field.rs
//! Demo of AUTO-SPLIT decomposition on a realistic irregular field.
//!
//! AutoSplit recursively bisects the field perpendicular to its longer
//! AABB side until every sub-field fits under `max_side`. Each sub-field
//! keeps its own headland rings (including along the shared split edge,
//! so there's proper turn space where swaths terminate at the split).
//!
//! A fixed fleet of N machines works EACH sub-field together (Block +
//! ByLength), then moves to the next. Each machine's tour is the
//! concatenation of its per-sub-field tours — per-part routing stays
//! intact, so no cross-field spaghetti.
//!
//! The strip between two sub-fields' innermost headland rings is the
//! shared midline turn band — drawn in cyan so you can see where
//! machines transition between halves.
//!
//! Run with:
//!   cargo run --example machine_split_field

#[path = "support/rerun_viz.rs"]
mod rerun_viz;

use maptrax::{
    Balance, DecompositionMode, DivisionPattern, DivisionPlan, FieldGenerationMode,
    FieldGenerationOptions, Geo, Maptrax, Point, RoutingOptions, RoutingStrategy, Swath,
    TurnPlannerConfig, TurnPlannerModel, point_xy, polygon_from_points, segment_length,
};
use rerun::Color;

fn big_irregular_field() -> maptrax::Polygon {
    polygon_from_points(vec![
        point_xy(-40.0, 90.0),
        point_xy(180.0, 0.0),
        point_xy(620.0, 30.0),
        point_xy(890.0, 180.0),
        point_xy(860.0, 420.0),
        point_xy(520.0, 495.0),
        point_xy(150.0, 450.0),
        point_xy(-60.0, 310.0),
    ])
}

const FLEET_SIZE: usize = 3;
const MAX_SIDE: f64 = 520.0;
const SWATH_WIDTH: f64 = 8.0;
const HEADLAND_COUNT: usize = 2;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec = rerun_viz::connect("maptrax_split_field")?;
    let border = big_irregular_field();

    let datum = Geo::new(51.0, 5.0, 0.0);

    rerun_viz::log_polygon(
        &rec,
        "enu/field/border",
        &border,
        Color::from_rgb(160, 100, 100),
    )?;
    rerun_viz::log_polygon_geo(
        &rec,
        "geo/field/border",
        &border,
        datum,
        Color::from_rgb(160, 100, 100),
    )?;

    let mut planner = Maptrax::new();
    planner.set_field(border.clone(), datum).expect("field");

    planner.plan_field(&FieldGenerationOptions {
        swath_width: SWATH_WIDTH,
        headland_count: HEADLAND_COUNT,
        decomposition: DecompositionMode::AutoSplit { max_side: MAX_SIDE },
        mode: FieldGenerationMode::ExplicitAngle(20.0),
    })?;

    let part_count = planner.field()?.get_parts().len();
    println!("=== AutoSplit {{ max_side: {MAX_SIDE} }} + {FLEET_SIZE}-machine fleet ===");
    println!("   Field extent: ~950m x ~495m (irregular 8-vertex polygon)");
    println!(
        "   Swath angle 20°, swath width {}m, {} headland rings per sub-field",
        SWATH_WIDTH, HEADLAND_COUNT
    );
    println!("   Split into {part_count} sub-field(s).");
    println!();

    // Per-sub-field headland rings. On an AutoSplit shared edge, the
    // non-owned side keeps its outermost headland on the split line (no empty
    // seam) and offsets deeper rings normally (no collapsed/overlapping
    // multi-headland lines).
    for (part_index, part) in planner.field()?.get_parts().iter().enumerate() {
        for (ring_index, ring) in part.headlands.iter().enumerate() {
            rerun_viz::log_polygon(
                &rec,
                &format!("enu/parts/part_{part_index}/headland_{ring_index}"),
                &ring.polygon,
                Color::from_rgb(120, 120, 160),
            )?;
            rerun_viz::log_polygon_geo(
                &rec,
                &format!("geo/parts/part_{part_index}/headland_{ring_index}"),
                &ring.polygon,
                datum,
                Color::from_rgb(120, 120, 160),
            )?;
        }
    }

    let plan = DivisionPlan::uniform(FLEET_SIZE, DivisionPattern::Block, Balance::ByLength);
    let turn = TurnPlannerConfig {
        swath_width: SWATH_WIDTH,
        min_turning_radius: 3.0,
        model: TurnPlannerModel::ReedsShepp,
        machine_length: 6.0,
        machine_width: 3.0,
        ..TurnPlannerConfig::default()
    };
    let routing = RoutingOptions {
        strategy: RoutingStrategy::GreedyNearest,
        local_improvement_passes: 1,
    };

    let all_plans = planner.plan_machines_for_all_parts(&plan, routing, &turn)?;

    // For each physical machine, stitch together the per-part tours in part
    // order. Each per-part tour was already routed inside that part with
    // its own headlands — no re-routing, no cross-field jumps.
    let mut assigned_per_machine: Vec<Vec<Swath>> = vec![Vec::new(); FLEET_SIZE];
    let mut tour_per_machine: Vec<Vec<Swath>> = vec![Vec::new(); FLEET_SIZE];
    let mut arcs_per_machine: Vec<Vec<Vec<Point>>> = vec![Vec::new(); FLEET_SIZE];

    for (part_index, planned) in all_plans.iter().enumerate() {
        for machine in &planned.machines {
            let id = machine.machine_index;
            assigned_per_machine[id].extend(machine.assigned_swaths.iter().cloned());
            tour_per_machine[id].extend(machine.tour.iter().cloned());
            arcs_per_machine[id].extend(machine.assigned_headland_arcs.iter().cloned());

            let color = machine_color(id);
            rerun_viz::log_swaths_tinted(
                &rec,
                &format!("enu/parts/part_{part_index}/swaths/machine_{id}"),
                &machine.assigned_swaths,
                color,
            )?;
            rerun_viz::log_swaths_geo_tinted(
                &rec,
                &format!("geo/parts/part_{part_index}/swaths/machine_{id}"),
                &machine.assigned_swaths,
                datum,
                Some(color),
            )?;
        }
    }

    println!(
        "   {:<12} {:>12} {:>12} {:>12}",
        "machine", "swath_count", "work_len_m", "work_s"
    );
    let mut makespan = 0.0f64;
    let mut total_work_len = 0.0;
    for machine_index in 0..FLEET_SIZE {
        let color = machine_color(machine_index);
        rerun_viz::log_swaths_tinted(
            &rec,
            &format!("enu/machines/m{machine_index}/swaths"),
            &assigned_per_machine[machine_index],
            color,
        )?;
        rerun_viz::log_polylines(
            &rec,
            &format!("enu/machines/m{machine_index}/headlands"),
            &arcs_per_machine[machine_index],
            color,
        )?;
        rerun_viz::log_swaths_tinted(
            &rec,
            &format!("enu/machines/m{machine_index}/tour"),
            &tour_per_machine[machine_index],
            color,
        )?;
        rerun_viz::log_swaths_geo_tinted(
            &rec,
            &format!("geo/machines/m{machine_index}/swaths"),
            &assigned_per_machine[machine_index],
            datum,
            Some(color),
        )?;
        rerun_viz::log_polylines_geo(
            &rec,
            &format!("geo/machines/m{machine_index}/headlands"),
            &arcs_per_machine[machine_index],
            datum,
            color,
        )?;
        rerun_viz::log_swaths_geo_tinted(
            &rec,
            &format!("geo/machines/m{machine_index}/tour"),
            &tour_per_machine[machine_index],
            datum,
            Some(color),
        )?;

        let work_len: f64 = assigned_per_machine[machine_index]
            .iter()
            .map(|s| segment_length(s.line))
            .sum();
        let work_s = work_len;
        println!(
            "   m{:<11} {:>12} {:>12.1} {:>12.1}",
            machine_index,
            assigned_per_machine[machine_index].len(),
            work_len,
            work_s,
        );
        makespan = makespan.max(work_s);
        total_work_len += work_len;
    }

    println!();
    println!("   Total work length: {:.1} m", total_work_len);
    println!(
        "   Makespan (slowest machine wins): {:.1} s   ~{:.0}% improvement vs single-machine baseline",
        makespan,
        (1.0 - makespan / total_work_len) * 100.0
    );
    println!(
        "   Each machine's tour = concat(part_0_tour, part_1_tour, …) — no re-routing between parts."
    );

    rec.flush_blocking()?;
    Ok(())
}

fn machine_color(machine_id: usize) -> (u8, u8, u8) {
    match machine_id {
        0 => (230, 90, 90),
        1 => (70, 180, 120),
        2 => (90, 140, 230),
        3 => (220, 170, 60),
        _ => (160, 160, 160),
    }
}
